### DISCUSSION TITLE
Можете, будь ласка, дати приклади конструкцій JS які не можуть бути написані в TS ?

### DISCUSSION THREAD

**Discussion by pasha-vuiko** [HowProgrammingWorks/LiveQA#32] [Q&A] [answered]
Можете, будь ласка, дати приклади конструкцій JS які не можуть бути написані в TS ?  Часто чув у ваших докладах що таких конструкцій чимало, поки знаю тільки про асинхронний конструктор, був би вдячний дізнатися побільше таких конструкцій =))

**Comment by tshemsedinov** [comment, marked as answer]
Все, що робить динамічне зв'язування типів та контрактів, а у JavaScript це робиться без проблем, усі перевірки у рантаймі. Наприклад:
- Фабрика функцій із невідомими контрактами (контракти можуть бути продифіновані у метаданих, структурах, що динамічно зчитуються за конфігів чи БД), це потрібно, коли ми робими інтроспекцію у дістрібьютид системі, де контракти могли змінитися і ми повинні зробити діскавері сервісу та контракту, а потім обернути їх у локальні виклики, щоб вони стали доступними з також динамічно завантеженого кода бізнес-логіки. Це метапрограмування, по якому у мене є лекції.
- Фабрики класів (використання таке ж, як у попередньому пункті)
- Фабрики прототипів (таке ж використання)

Тобто, взагалі, можна їх трактувати як щось `unknown`, але у чому тоді сенс ts?

Як з цим бути? Писати `.d.ts` окремо від `.js` коду та динамічно перевіряти усі ці контракти у рантаймі через схеми, а якщо у дизайнтаймі треба, то генерувати `.d.ts` зі схем.

**Reply by pasha-vuiko** [reply]
Дякую)
